nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java

```diff
@@ -0,0 +1,293 @@
+
+        FlowFile flowFile = null;
+        if (context.hasIncomingConnection()) {
+            flowFile = session.get();
+
+            // If we have no FlowFile, and all incoming connections are self-loops then we can continue on.
+            // However, if we have no FlowFile and we have connections coming from other Processors, then
+            // we know that we should run only if we have a FlowFile.
+            if (flowFile == null && context.hasNonLoopConnection()) {
+                return;
+            }
+        }
+
+        OkHttpClient okHttpClient = getClient();
+
+        if (flowFile == null) {
+            flowFile = session.create();
+        }
+
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions(flowFile).getValue();
+        final String docId = context.getProperty(DOC_ID).evaluateAttributeExpressions(flowFile).getValue();
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(flowFile).getValue();
+        final String fields = context.getProperty(FIELDS).isSet()
+                ? context.getProperty(FIELDS).evaluateAttributeExpressions(flowFile).getValue()
+                : null;
+
+        final ComponentLog logger = getLogger();
+
+
+        try {
+            logger.debug("Fetching {}/{}/{} from Elasticsearch", new Object[]{index, docType, docId});
+
+            // read the url property from the context
+            final String urlstr = StringUtils.trimToEmpty(context.getProperty(ES_URL).getValue());
+            final URL url = buildRequestURL(urlstr, docId, index, docType, fields);
+            final long startNanos = System.nanoTime();
+
+            final Response getResponse = sendRequestToElasticsearch(okHttpClient, url, "GET", null);
+            final int statusCode = getResponse.code();
+
+            if (isSuccess(statusCode)) {
+                ResponseBody body = getResponse.body();
+                final byte[] bodyBytes = body.bytes();
+                JsonNode responseJson = parseJsonResponse(new ByteArrayInputStream(bodyBytes));
+                boolean found = responseJson.get("found").asBoolean(false);
+
+                if (found) {
+                    flowFile = session.putAttribute(flowFile, "filename", docId);
+                    flowFile = session.putAttribute(flowFile, "es.index", index);
+                    flowFile = session.putAttribute(flowFile, "es.type", docType);
+                    flowFile = session.write(flowFile, out -> {
+                        out.write(bodyBytes);
+                    });
+                    logger.debug("Elasticsearch document " + docId + " fetched, routing to success");
+
+                    // emit provenance event
+                    final long millis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                    session.getProvenanceReporter().fetch(flowFile, url.toExternalForm(), millis);
```

Comment:
"Fetch" provenance events should be emitted when there was an incoming FlowFile that was used to fetch the information. "Receive" events should be emitted when the processor is a source processor and there was no incoming flowfile. There should be some logic to determine which case this falls into.